azure/Guidelines.md

```diff
@@ -0,0 +1,873 @@
+ 
+> :no_entry: **DO NOT** adopt this pattern. If you feel you need an exception, contact the Azure HTTP/REST Stewardship Board <b>prior</b> to implementation.
+ 
+*If you feel you need an exception, or need clarity based on your situation, please contact the Azure HTTP/REST Stewardship Board <b>prior</b> to release of your API.*
+
+## Building Blocks: HTTP, REST, & JSON
+The Microsoft Azure Cloud platform exposes its APIs through the core building blocks of the Internet; namely HTTP, REST, and JSON. This section provides you with a general understanding of how these technologies should be applied when creating your service.
+
+### HTTP
+Azure services must adhere to the HTTP specification, [RFC7231](https://tools.ietf.org/html/rfc7231). This section further refines and constrains how service implementors should apply the constructs defined in the HTTP specification. It is therefore, important that you have a firm understanding of the following concepts:
+* [Uniform Resource Locators (URLs)](URLS)
+* HTTP Methods
+* Request & Response Headers
+* Bodies
+
+### Uniform Resource Locators (URLs)
+
+A Uniform Resource Locator (URL) is how developers access the resources of your service. Ultimately, URLs are how developers form a cognitive model of your service's resources.
+
+> :white_check_mark: **DO** use this URL pattern:
+```text
+https://<service>.<cloud>/<tenant>/<service-root>/<resource-collection>/<resource-id>/
+```
+
+Where:
+ | Field | Description
+ | - | - |
+ | service | Name of the service (ex: blobstore, servicebus, directory, or management)
+ | cloud | Cloud domain name (see Azure CLI's "az cloud list")<p><table><tr><td><b>Cloud</td><td><b>Domain</td></tr><tr><td>Public</td><td>azure.net</td></tr></table>
+ | tenant | Globally-unique ID of container representing tenant isolation, billing, enforced quotas, lifetime of containees (ex: subscription UUID)
+ | service&#x2011;root | Service-specific path (ex: blobcontainer, myqueue)
+ | resource&#x2011;collection | Name of the collection, unabbreviated, pluralized
+ | resource&#x2011;id | Value of the unique id property. This MUST be the raw string/number/guid value with no quoting but properly escaped to fit in a URL segment.
+
+> :white_check_mark: **DO** use kebab-casing (preferred) or camel-casing for URL path segments. If the segment refers to a JSON field, use camel casing.
+> 
+> :white_check_mark: **DO** return ```414-URI Too Long``` if a URL exceeds 2083 characters
+>
+> :white_check_mark: **DO** treat URL path segments as case-sensitive. If the passed-in case doesn't match what the service expects, the request __MUST__ fail with a ```404-Not found``` HTTP return code.
+
+> :ballot_box_with_check: **YOU SHOULD** keep URLs readable; if possible, avoid UUIDs & %-encoding (ex: Cádiz is %-encoded as C%C3%A1diz)
+> 
+> :ballot_box_with_check: **YOU SHOULD** limit your URL's characters to `0-9  A-Z  a-z  -  .  _  ~ :`
```

Comment:
If we are recommending either kebab-case or camelCase for path segments (see line 69), then I think we should move the “_” character from the “YOU SHOULD” to the “YOU MAY” list of URL characters.

Reply:
The characters that never need %-encoding are 0-9  A-Z  a-z  -  .  _  ~; not colon. So I think colon should be removed from the list. 

If you want: there is a difference between service-defined path segments and customer-specified path segments. I think the service should stick with 0-9  A-Z  a-z  -  .  _  ~ (so make this a DO guideline; not YOU SHOULD). We reserve colon for actions only (so don't list it here so that service teams avoid it). 

Customers can use 0-9  A-Z  a-z  -  .  _  ~ and other %-encoded values. They should avoid colon and other special characters. For example, what if the customer uses quote? Then we might return a json field with "id":"database"Name" (quote between database & name on purpose) - this is not valid JSON.
Or, if the customer uses ?, then the path ends and the query parameters begin - bad.
Or, if they use /, then they are inserting a path segment - bad.
Or, if they use #, then they are starting a fragment - bad.
Or, if they use +, then sometimes its interpreted as a space - bad.

Service teams don't think about any of this so it would be good to be more explicit in these guidelines for service vs customer characters. 
Customers segments MAY support the following characters (or a subset thereof): 0-9  A-Z  a-z  -  .  _  ~  !  $  &   '  (  )  *  ,   ;  =  @  [] (space). NOTE: I took a stab at this list but am open to additions/removals.